File: java/leetcode_grind/Day640.java
package leetcode_grind;

import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Set;
import java.util.TreeSet;

public class Day640 {
    // https://leetcode.com/problems/ugly-number-ii/description/?envType=daily-question&envId=2024-08-18
    static class Solution1 {
        public int nthUglyNumber(int n) {
            TreeSet<Long> uglyNumbersSet = new TreeSet<>();
            uglyNumbersSet.add(1L);
            Long currentUgly = 1L;
            for (int i = 0; i < n; i++) {
                currentUgly = uglyNumbersSet.pollFirst();
                uglyNumbersSet.add(currentUgly * 2);
                uglyNumbersSet.add(currentUgly * 3);
                uglyNumbersSet.add(currentUgly * 5);
            }
            return currentUgly.intValue();
        }
    }

    static class Solution2 {
        public int nthUglyNumber(int n) {
            PriorityQueue<Long> minHeap = new PriorityQueue<>();
            Set<Long> seenNumbers = new HashSet<>();
            int[] primeFactors = { 2, 3, 5 };
            minHeap.offer(1L);
            seenNumbers.add(1L);
            long currentUgly = 1L;
            for (int i = 0; i < n; i++) {
                currentUgly = minHeap.poll();

                for (int prime : primeFactors) {
                    long nextUgly = currentUgly * prime;
                    if (seenNumbers.add(nextUgly)) {
                        minHeap.offer(nextUgly);
                    }
                }
            }
            return (int) currentUgly;
        }
    }
}

File: src/leetcode_grind/day_640.rs
// https://leetcode.com/problems/ugly-number-ii/description/?envType=daily-question&envId=2024-08-18
pub fn nth_ugly_number(n: i32) -> i32 {
    let mut ugly_numbers = vec![0; n as usize];
    ugly_numbers[0] = 1;

    let mut index_multiple_of_2 = 0;
    let mut index_multiple_of_3 = 0;
    let mut index_multiple_of_5 = 0;
    
    let mut next_multiple_of_2 = 2;
    let mut next_multiple_of_3 = 3;
    let mut next_multiple_of_5 = 5;

    for i in 1..n as usize {
        let next_ugly_number = next_multiple_of_2.min(
            next_multiple_of_3.min(
                next_multiple_of_5
            ) 
        );

        ugly_numbers[i] =  next_ugly_number;

        if next_ugly_number == next_multiple_of_2 {
            index_multiple_of_2 += 1;
            next_multiple_of_2 = ugly_numbers[index_multiple_of_2] * 2;
        }

        if next_ugly_number == next_multiple_of_3 {
            index_multiple_of_3 += 1;
            next_multiple_of_3 = ugly_numbers[index_multiple_of_3] * 3;
        }

        if next_ugly_number == next_multiple_of_5 {
            index_multiple_of_5 += 1;
            next_multiple_of_5 = ugly_numbers[index_multiple_of_5] * 5;
        }

    }

    ugly_numbers[n as usize - 1]
}